tools/src/main/java/org/apache/kafka/tools/CompatibilityTest.java

```diff
@@ -0,0 +1,359 @@
+                    @Override
+                    public void run() {
+                        offsetsForTime.result = consumer.offsetsForTimes(timestampsToSearch);
+                    }
+                },
+                new Runnable() {
+                    @Override
+                    public void run() {
+                        log.info("offsetsForTime = {}", offsetsForTime.result);
+                    }
+                });
+        // Whether or not offsetsForTimes works, beginningOffsets and endOffsets
+        // should work.
+        consumer.beginningOffsets(timestampsToSearch.keySet());
+        consumer.endOffsets(timestampsToSearch.keySet());
+
+        consumer.assign(topicPartitions);
+        consumer.seekToBeginning(topicPartitions);
+        final Iterator<byte[]> iter = new Iterator<byte[]>() {
+            private final int timeoutMs = 10000;
+            private Iterator<ConsumerRecord<byte[], byte[]>> recordIter = null;
+            private byte[] next = null;
+
+            private byte[] fetchNext() {
+                while (true) {
+                    long curTime = Time.SYSTEM.milliseconds();
+                    if (curTime - prodTimeMs > timeoutMs)
+                        throw new RuntimeException("Timed out after " + timeoutMs + " ms.");
+                    if (recordIter == null) {
+                        ConsumerRecords<byte[], byte[]> records = consumer.poll(100);
+                        recordIter = records.iterator();
+                    }
+                    if (recordIter.hasNext())
+                        return recordIter.next().value();
+                    recordIter = null;
+                }
+            }
+
+            @Override
+            public boolean hasNext() {
+                if (next != null)
+                    return true;
+                next = fetchNext();
+                return next != null;
+            }
+
+            @Override
+            public byte[] next() {
+                if (!hasNext())
+                    throw new NoSuchElementException();
+                byte[] cur = next;
+                next = null;
+                return cur;
+            }
+
+            @Override
+            public void remove() {
+                throw new UnsupportedOperationException();
+            }
+        };
+
+        byte[] next = iter.next();
+        try {
+            compareArrays(message1, next);
+            log.debug("Found first message...");
+        } catch (RuntimeException e) {
+            throw new RuntimeException("The first message in this topic was not ours.  Please use a new " +
+                "topic when running this program.\n");
+        }
+        next = iter.next();
+        try {
+            compareArrays(message2, next);
+            log.debug("Found second message...");
+        } catch (RuntimeException e) {
+            throw new RuntimeException("The first message in this topic was not ours.  Please use a new " +
```

Comment:
I think you meant `second` message here. Why not simply accumulate the results into a collection of Strings and then do `arrayEquals(List("foo1", "foo2"), messages)` or something like that? It seems like the code would be simpler.